[torch.export] Support is_compiling() flag for non-strict mode

khabinov opened the pull request:
Summary: In non-strict mode of torch.export() we didn't set those `is_compiling()` to `True` which is needed by some models.

Test Plan: Unit tests and manual testing.

Differential Revision: D53624452




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

khabinov commented on torch/compiler/__init__.py:
```diff
@@ -149,3 +150,23 @@ def wrap_numpy(fn):
+
+_is_compiling_flag = False
+
+def is_compiling() -> bool:
+    """
+    Indicates whether a graph is executed as part of torch.compile() or torch.export().
+
+    Note that there are 2 other related flags (we should probably deprecate them eventually):
+      * torch._dynamo.external_utils.is_compiling() https://github.com/pytorch/pytorch/blob/a7f82b7d628eb2b966bc53e593dcf32049b2b10e/torch/_dynamo/external_utils.py#L14
+      * torch._utils.is_compiling() https://github.com/pytorch/pytorch/blob/a7f82b7d628eb2b966bc53e593dcf32049b2b10e/torch/_utils.py#L852
+
+    Example::
+
+        >>> def forward(self, x):
+        >>>     if not torch.compiler.is_compiling():
+        >>>        ...logic that is not needed in a compiled/traced graph...
+        >>>
+        >>>     ...rest of the function...
+    """
+    return _is_compiling_flag
```
Given that there are already 2 other flags in the code base, I am not sure about increased complexity of having 3 flags doing the same thing :)

cc @suo, @ezyang, @mlazos

suo replied:
The other ones are technically non-public APIs, so I think we should redirect them to the public one and slowly try to extinguish them :P

mlazos replied:
I think you should be able to replace torch._utils.is_compiling() with the torch._dynamo.eternal_utils.is_compiling() right now.

khabinov replied:
> I think you should be able to replace torch._utils.is_compiling() with the torch._dynamo.eternal_utils.is_compiling() right now.

Thanks for reviewing! Yeah, I can do this as part of this diff.

suo commented:
lgtm—do we want to block on rerouting existing APIs to use this one as a source of truth? will let @mlazos make the call

mlazos commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -6,6 +6,7 @@
 from torch._higher_order_ops.utils import (
     _has_potential_branch_input_alias,
     _has_potential_branch_input_mutation,
+    _is_compiling_hoo,
```
should this be hook?

khabinov replied:
Should be `hoo` :) It stands for "higher-order op", I saw this abbreviation used here and there.

mlazos commented on torch/_higher_order_ops/utils.py:
```diff
@@ -76,15 +76,29 @@ def graph_with_interpreter(*args):
+_is_compiling_hoo_flag: bool = False
+
+
+def _is_compiling_hoo() -> bool:
```
hmm did we need this due to circular imports?

khabinov replied:
Actually, this is a part of my change that I am not 100% sure about. I've noticed a recurring pattern in the HOO code  like:

```python
if torch._dynamo.is_compiling():
    return some_higher_order_op()

<...>
return torch.compile(some_higher_order_op, ...)
```

Since in this PR I made sure to set `is_compiling` flag to `True`, the code inside `if`-block started to get executed instead of going inside `torch.compile()`, so I thought that probably a separate flag should be created for HOO case.

cc @tugsbayasgalan, @suo

khabinov replied:
> torch._dynamo.is_compiling() returning True in this case is the existing behavior right?

Yeah, previously it would return `False` for non-strict mode, thus we would never go inside the `if`-block. Now for non-strict mode it returns `True`, so we get inside the `if`-block and fail with `FunctionalTensorMode`-related error.

I don't fully understand this `if`-block, tried to find unit tests that would cover it, but didn't find anything, therefore any opinion from HOO experts would be appreciated.

mlazos replied:
cc @zou3519 do you know about this?

sijiac replied:
n00q, what's does the "HOO" mean? I didn't find any clue related to it, thanks

suo replied:
I don't think a separate `is_compiling_hoo()` makes sense.

Looking at the `cond` example, I believe the desired behavior is for us to go down the `True` path in non-strict mode, where we will call the higher-order op and then drop into the ProxyTensorMode implementation.

But it appears you have tripped a latent issue, which is that `torch.cond` and friends don't work with non-strict mode. So we probably should address that directly (not in this PR).

For now, let's just disable any non-strict tests that fail with the new behavior. This not a regression, since they were silently broken anyway. 

@ydwu4, please double-check my reasoning here + can you follow-up on the control flow support in non-strict?

suo replied:
Okay, that means I was wrong, and that the desired behavior (at least today) is that we go down the `False` branch in non-strict mode. Which means some version of this is necessary.

@khabinov: my recommendation is that instead of a special `is_compiling_hoo`, we make a more general-purpose `is_dynamo_compiling` flag, which is true only if we are in dynamo (basically the old behavior of `torch._dynamo.is_compiling`), and we use it here specifically. That would communicate the intention of this code better.

khabinov committed:
[torch.export] Support is_compiling() flag for non-strict mode (#119602)

Summary:

In non-strict mode of torch.export() we didn't set those `is_compiling()` to `True` which is needed by some models.

Test Plan: Unit tests and manual testing.

Reviewed By: suo

Differential Revision: D53624452
